[quant][observer] Add histogram observer

hx89 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24022 [quant][observer] Add l2 norm minimization
* **#23959 [quant][observer] Add histogram observer**

Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
[quant][observer] Add histogram observer

Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -103,6 +103,127 @@ def calculate_qparams(self, **kwargs):
+            new_min = torch.min(x)
+            new_max = torch.max(x)
+            new_histogram = torch.histc(x, self.bins)
+            # combine the existing histogram and new histogram into 1 histogram
```
For the first version of histogram observer, suggest having a fixed width histogram, so that combining histograms is trivial.
```python
 def forward(self, x):
        if self.min_val is None or self.max_val is None or self.histogram is None:
            self.min_val = torch.min(x)
            self.max_val = torch.max(x)
            range = self.max_val -self.min_val
            self.relaxed_min = self.min_val - 0.5*range
            self.relaxed_max = self.max_val + 0.5*range
            self.histogram = torch.histc(input, self.bins, min=self.relaxed_min, max=self.relaxed_max)
        else:
            new_min = torch.min(x)
            new_max = torch.max(x)
            new_histogram = torch.histc(x, self.bins, min=relaxed_min, max=relaxed_max)
            self. histogram = new_histogram+self.histogram
```

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -103,6 +103,127 @@ def calculate_qparams(self, **kwargs):
+    def calculate_qparams(self, **kwargs):
+        if self.histogram is None:
+            raise Exception("must run observer before calling calculate_qparams!")
+        min_bin = 0
```
Can do this with torch functions as:
```python
# y is the input histogram
histogram_mask = torch.gt(y,0).type(torch.int8)
c = torch.cumsum(histogram_mask,0)
# Last non-zero bin
max_bin = torch.argmax(histogram_mask)
# Only one entry is non-zero, find it.
min_bin = torch.argmax(torch.eq(c,1).type(torch.int8))
```

raghuramank100 reviewed (changes requested):
Please see comments, thanks

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

hx89 commented:
Updated to address comments.

hx89 commented:
@raghuramank100 , I have addressed your comments, could you please take a look?

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

llyfacebook commented on torch/quantization/observer.py:
```diff
@@ -120,14 +126,80 @@ def calculate_qparams(self):
+            range = max_val - min_val
+            self.relaxed_min = min_val - 0.5 * range
+            self.relaxed_max = max_val + 0.5 * range
+            self.histogram = torch.histc(x, self.bins, min=self.relaxed_min, max=self.relaxed_max)
```
It seems I trigged one issue for self.relaxed_min. I could put the convert(model) function under with torch.no_grad(): to work around the problem. 
`TypeError: histc(): argument 'min' must be Number, not Tensor`

hx89 replied:
@llyfacebook, I've updated the PR to use self.relaxed_min.item() inside torch.histc(), which should resolve the issue. Could you let me know how to reproduce the TypeError so that I can double check?

llyfacebook replied:
Just not put your convert(model) within "with torch.no_grad()", it should trigger the issue.

hx89 replied:
I've tested the new PR works without error, thanks for your feedback @llyfacebook !

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -137,16 +144,80 @@ def calculate_qparams(self):
+            self.min_val = self.relaxed_min
+            self.max_val = self.relaxed_max
+        else:
+            new_histogram = torch.histc(x, self.bins, min=self.relaxed_min.item(), max=self.relaxed_max.item())
```
Please add a warning if the min/max of the incoming data in subsequent batches is outside the relaxed_min/max range.

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -137,16 +144,80 @@ def calculate_qparams(self):
+            new_min = min_val + min_bin * bin_width
+            new_max = min_val + (max_bin + 1) * bin_width
+            return self._calculate_qparams(new_min, new_max)
+
```
Should we have a different repr() for histogram observer? Histograms can have too many bins, so not sure what exactly we could do here.

hx89 replied:
I may not be familiar with the terminology, what is repr()?

raghuramank100 replied:
repr() is how the class shows up when you do print(observer())

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -765,5 +767,26 @@ def test_observer_scriptable(self):
         self.assertEqual(obs.calculate_qparams(), loaded.calculate_qparams())
 
+    @given(qdtype=st.sampled_from((torch.qint8, torch.quint8)),
+           qscheme=st.sampled_from((torch.per_tensor_affine, torch.per_tensor_symmetric)))
```
The new test code also checks for reduce_range, please rebase and merge into that

raghuramank100 reviewed (changes requested):
Looks good, please address comments so that is ready to go

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

hx89 commented:
@raghuramank100 , I've addressed your comments, could you please take a look?

dzhulgakov reviewed (commented):
can you just merge it with the next PR in the stack? they seem to be doing the same change

hx89 commented:
@dzhulgakov These two PR are different even though their changes are both within histogram observer, this PR adds the basic histogram observer with fixed width and compute the qparams based on histogram range, and the next PR uses more advanced way to merge histograms and compute qparams using l2 norm minimization for better accuracy. Maybe it's better to keep them separate to keep each PR more focused. I can add options in the next PR for users to choose between different ways to combine histogram and compute qparams.

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

Commit:
Update on "[quant][observer] Add histogram observer"


Add histogram observer that records the running histogram of tensor values along with min/max values.

Differential Revision: [D16692835](https://our.internmc.facebook.com/intern/diff/D16692835/)

smessmer commented:
Reverted in https://github.com/pytorch/pytorch/pull/26236 because it broke CI
